OutOfSchool/Tests/OutOfSchool.WebApi.IntegrationTests/ProviderServiceIntergrationTests/ProviderServiceUpdate.cs

```diff
@@ -0,0 +1,241 @@
+            usersRepositoryMock = CreateUsersRepositoryMock(fakeUser);
+
+            ratingService = new Mock<IRatingService>();
+            localizer = new Mock<IStringLocalizer<SharedResource>>();
+            logger = new Mock<ILogger<ProviderService>>();
+            adkressRepository = new Mock<IEntityRepository<Address>>();
+            providerRepository = new ProviderRepository(GetContext);
+            providerService = new ProviderService(providerRepository, usersRepositoryMock.Object, ratingService.Object,
+                logger.Object, localizer.Object, mapper, adkressRepository.Object);
+        }
+
+        [Test]
+        public async Task UpdateWhenProviderHasSameAdresses_WithSameAddresses_UpdatesOneAddress()
+        {
+            // Arrange
+            await using var context = GetContext;
+            var provider = context.Providers.First();
+            provider.ActualAddressId = null;
+            await context.SaveChangesAsync().ConfigureAwait(false);
+            provider.ActualAddress = provider.LegalAddress;
+
+            // Act
+            var result = await providerService.Update(mapper.Map<ProviderDto>(provider), provider.UserId, "Provider")
+                .ConfigureAwait(false);
+
+            // Assert
+            Assert.IsFalse(result.ActualAddress.IsSameOrEqualTo(result.LegalAddress));
+            Assert.IsNotNull(result.LegalAddress);
+            Assert.IsNull(result.ActualAddress);
+        }
+
+        [Test]
+        public async Task UpdateWhenProviderHasSameAdresses_WithActualAddressNull_UpdatesOneAddress()
+        {
+            // Arrange
+            await using var context = GetContext;
+            var provider = context.Providers.First();
+            provider.ActualAddressId = null;
+            await context.SaveChangesAsync().ConfigureAwait(false);
+            var providerDto = mapper.Map<ProviderDto>(provider);
+            providerDto.ActualAddress = null;
+
+            // Act
+            var result = await providerService.Update(providerDto, provider.UserId, "Provider")
+                .ConfigureAwait(false);
+
+            // Assert
+            Assert.IsFalse(result.ActualAddress.IsSameOrEqualTo(result.LegalAddress));
+            Assert.IsNotNull(result.LegalAddress);
+            Assert.IsNull(result.ActualAddress);
+        }
+
+        [Test]
+        public async Task UpdateWhenProviderHasSameAdresses_WithNewLegalAddressAndActualIsPreviousLegal_UpdatesOneAddress()
+        {
+            // Arrange
+            await using var context = GetContext;
+            var provider = context.Providers.First();
+            provider.ActualAddressId = null;
+            await context.SaveChangesAsync().ConfigureAwait(false);
+            await context.Entry(provider).ReloadAsync().ConfigureAwait(false);
+            var randomAddress = AddressGenerator.Generate();
+            randomAddress.Id = 0;
```

Comment:
```
var randomAddress = AddressGenerator.Generate();
randomAddress.Id = 0;
```
mentioned pattern is duplicated in code. 
1. it looks like `randomAddress` should be named `randomAddressToAdd` (because `Id = 0` means that it is not taken from db and thus should be added)
2. You could provide small helper method to reduce code dublication